neuralmonkey/encoders/cnn_encoder.py

```diff
@@ -139,17 +255,111 @@ def output(self) -> tf.Tensor:
             dropout_keep_prob=self.dropout_keep_prob,
             train_mode=self.train_mode)
 
-    @tensor
-    def spatial_mask(self) -> tf.Tensor:
-        return tf.ones(tf.shape(self.spatial_states)[:3])
-
     def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
         # if it is from the pickled file, it is list, not numpy tensor,
         # so convert it as as a prevention
         images = np.array(dataset.get_series(self.data_id))
 
         f_dict = {}
         f_dict[self.image_input] = images / 225.0
+        # it is one everywhere where non-zero, i.e. zero columns are masked out
+        f_dict[self.image_mask] = np.sum(
+            np.sign(images), axis=3, keepdims=True)
 
         f_dict[self.train_mode] = train
         return f_dict
+
+
+def max_pooling(
+        prev_layer: tf.Tensor,
+        prev_mask: tf.Tensor,
+        specification: Tuple[str, int, int, str],
+        layer_num: int) -> Tuple[tf.Tensor, tf.Tensor]:
+    if len(specification) != 4:
+        raise ValueError((
+            "Specification of a convolutional layer needs to have 5 "
+            "members: kernel size, stride, padding, was {}")
+                         .format(specification))
+    pool_size, stride, pad = specification[1:]
+
+    if pad not in ["same", "valid"]:
+        raise ValueError(
+            "Padding must be 'same' or 'valid', was '{}' in layer {}."
+            .format(pad, layer_num + 1))
+
+    with tf.variable_scope("layer_{}_max_pool".format(layer_num)):
+        next_layer = tf.layers.max_pooling2d(prev_layer, pool_size, stride)
+        next_mask = tf.layers.max_pooling2d(prev_mask, pool_size, stride)
+    return next_layer, next_mask
+
+
+class OCREncoder(ModelPart, TemporalStatefulWithOutput):
```

Comment:
v týhle třídě je okopírovaná spousta kódu, nedalo by se to nějak podědit?

Reply:
Správný řešení by as bylo udělat na CNN wrapper, kterej to nakrájí jako `TemporalStateful` a na to už volat normální enkodéry. Psal jsem to ale ještě dřív, než na to byly ty enkodéry připravé, a už mám s tímhle natrénované modely.

Reply:
takže navrhuješ tohle nechat takhle a časem to někdy překopat?

Reply:
Jo.

Reply:
Změna plánu, natrénovaný modely nechci, protože jsou zatížený bugem v načítání obrázků, takže to udělám pořádně.

Reply:
Cool. každej bug je na něco dobrej. :-)